### PR TITLE
TASK-12: discovery page tabs

### PR DESCRIPTION
**Pull request by r-saba**
This PR implements the discovery page tabs #54.

To test the component visit the followin URL:
`http://localhost:4200/home`

Swiping between routes is possible when in mobile view

### CODE REVIEW

File: package.json
Comment (cristian-aldea) [inline package.json:72]: I'm getting the following warning in the console, which seems to be coming from the ReactSwipableView:
```
Warning: Legacy context API has been detected within a strict-mode tree.

The old API will be supported in all 16.x releases, but applications using it should migrate to the new version.

Please update the following components: ReactSwipableView

Learn more about this warning here: https://fb.me/react-legacy-context
    in ReactSwipableView (created by SwipeableRoutes)
    in SwipeableRoutes (created by Context.Consumer)
    in withRouter(SwipeableRoutes) (created by App)
    in Route (created by App)
    in Switch (created by App)
    in div (created by App)
    in App
    in Router (created by BrowserRouter)
    in BrowserRouter
    in ApolloProvider
    in StrictMode
printWarning @ vendor.js:127606
vendor.js:127606 

Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.

Please update the following components: ReactSwipableView
```

Would there be a way for us to implement this sliding ourselves? Or update the package to not have this warning?

File: apps/mull-ui/src/app/app.scss
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.scss]: This could be moved to `libs/ui-lib/src/lib/styles/styles.scss`

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx]: I think it would be better if we kept the home page and made the three pages you added be subpages, so something like:
- `/home/discover`
- `/home/upcoming`
- `/home/myevents`

What do you think @r-saba?

File: apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx]:
```suggestion
import { NavLink } from 'react-router-dom';
```
Make sure to lint your code before making a pr. There are other linting warnings/errors I found from your changes, so please fix them.

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx]: I don't know about you @r-saba but I really don't like how the sub-navbar slides along with the page. It makes it really confusing. I experimented a bit and found a solution. Let me know what you think.
```suggestion
        <Route exact path={[`${ROUTES.DISCOVER}`, `${ROUTES.UPCOMING}`, `${ROUTES.MYEVENTS}`]}>
          <SubNavigationBar />
          <SwipeableRoutes>
            <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
            <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
            <Route path={ROUTES.MYEVENTS} component={() => <div>MYEVENTS!</div>} />
          </SwipeableRoutes>
        </Route>
```

**Review by cristian-aldea — CHANGES_REQUESTED** [review]
Nice implementation @r-saba! I had a few concerns and comments I'd like some clarifications on, but otherwise nice job.

File: apps/mull-ui/src/app/app.scss
Comment (r-saba) [reply apps/mull-ui/src/app/app.scss]: I put it here because we're getting rid of ui-lib but yeah I can put it there for now and we can move it all at once

File: apps/mull-ui/src/app/app.tsx
Comment (r-saba) [reply apps/mull-ui/src/app/app.tsx]: That's how I had it, initially I preferred it without the `/home`  since it with concise

File: apps/mull-ui/src/app/app.tsx
Comment (r-saba) [reply apps/mull-ui/src/app/app.tsx]: Oh good point, didn't consider that we probably don't want the subnavigation to slide as well

File: apps/mull-ui/src/app/app.scss
Comment (cristian-aldea) [reply apps/mull-ui/src/app/app.scss]: yeah for now keep it in there, we'll move it later

File: package.json
Comment (r-saba) [reply package.json:72]: This issue is currently present in almost all UI Framworks ([materia-ui](https://github.com/mui-org/material-ui/issues/13394), [react-strap](https://github.com/reactstrap/reactstrap/issues/1833)). Users have not reported any issues as a result of this warning.

Also [react-swipeable-views](https://github.com/oliviertassinari/react-swipeable-views) is widely used so I wouldn't worry about support for this package. It's used by Uber, Tinder and Material-UI. The warning in [known](https://github.com/oliviertassinari/react-swipeable-views/issues/596).

Considering all of this I don't see the benefit of making our own implementation.

Comment (cristian-aldea) [reply package.json:72]: Alright makes sense, just wanted a bit more insight into this.

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [reply apps/mull-ui/src/app/app.tsx]: considering we have a button for the home page id say the way you had it initially was good. Having them under a common base route of `/home` makes it easier to understand that its still the home page, but just different tabs.

Comment (cristian-aldea) [reply apps/mull-ui/src/app/app.tsx]: I don't mind either way, so I won't block over it.

**Review by cristian-aldea — COMMENTED** [review]
Thanks for addressing my comments, I don't have time to re-review, and I don't want to block this pr, so I'll approve it immediately if it gets another approval. I'll be re-reviewing in more detail tomorrow.

File: apps/mull-ui/src/constants.ts
Comment (bit172) [inline apps/mull-ui/src/constants.ts:16]: This is really minor but I think it should be consistent
```suggestion
  MY_EVENTS: '/my-events',
```

Comment (bit172) [inline apps/mull-ui/src/constants.ts]: This is really minor but I think it should be consistent
```suggestion
  MY_EVENTS: '/my-events',
```

**Review by bit172 — CHANGES_REQUESTED** [review]
Looks very good! I just have a really minor change. Good job!

**Review by JimLau49 — DISMISSED** [review]
Looks good to me 🚀  . I think it should have been mentioned that to test scrollability that you need to click along the length of container that has "DISCOVER!" "UPCOMING!" or "MYEVENTS!"
**Acceptance Criteria**

- [x] The tabs are implemented for the discovery page and the user can swipe/tap to switch between tabs.
- [x] Each tab is routed, but can lead to dead links or pages
- [x] Documentation present for every new component and function
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=533%3A6528)

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx]: Is there a reason why we can't have:
```suggestion
        <Route exact path={[ROUTES.DISCOVER, ROUTES.UPCOMING, ROUTES.MYEVENTS]}>
```
I tried it and it seems to work

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx]: One issue im seeing right now with styling is that the Subnavbar is set to be in the page-container, which is then pushing the actual page content further down than it should.

image

I propose this solution: Remove the page-container class from SubNavigationBar, and put page-container on a container div. So something like this:
```suggestion
        <Route exact path={[ROUTES.DISCOVER, ROUTES.UPCOMING, ROUTES.MYEVENTS]}>
          <div className="page-container">
            <SubNavigationBar />
            <SwipeableRoutes>
              <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
              <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
              <Route path={ROUTES.MYEVENTS} component={() => <div>MYEVENTS!</div>} />
            </SwipeableRoutes>
          </div>
        </Route>
```

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx]: The more I think about it the more I think these routes should be under `/home`, since each tab is a subpage for the home page. I'm not gonna block the PR over this, but I really think what you did initially was good.

File: apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx:7]: Add this to the `apps/mull-ui/src/app/components/index.ts` 👀

File: apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.spec.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.spec.tsx]: Same comment as in app.tsx, is there a reason we can't put:
```suggestion
      [ROUTES.UPCOMING]: 'subnavigation-upcoming-button',
      [ROUTES.MYEVENTS]: 'subnavigation-myEvents-button',
```
I tried it and the tests still seem to work.

**Review by cristian-aldea — CHANGES_REQUESTED** [review]
I had a chance to look at your pr in more detail and I found a few more issues I'd like addressed or replied to.

File: apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx
Comment (r-saba) [reply apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx:7]: you do a good job of catching these minor details!

File: apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.spec.tsx
Comment (r-saba) [reply apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.spec.tsx]: It's a remnant from my first iteration, I did a simple find and replace, should be what you proposed now.

File: apps/mull-ui/src/app/app.tsx
Comment (r-saba) [reply apps/mull-ui/src/app/app.tsx]: Sure, I will change it

File: apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx
Comment (cristian-aldea) [reply apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx:7]: More 👀 more 💵

**Comment by r-saba** [comment]
I addressed your pr comments @bit172 @TheGreatMarkus @JimLau49 . Let me know how it's looking.
Major difference is now each tab points to a nested route `/home/:id`

File: apps/mull-ui/src/app/app.tsx
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx:57]: I got one last suggestion here. This is to avoid people routing to /home/somethingelse
```suggestion
              <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
              <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
              <Route path={ROUTES.MY_EVENTS} component={() => <div>MY_EVENTS!</div>} />
              <Route path={ROUTES.HOME} render={() => <Redirect to={ROUTES.DISCOVER} />} />
```

**Review by cristian-aldea — DISMISSED** [review]
Great Work Ragith! I just had one last issue I saw. But otherwise this is as good as merged.

- [x] The tabs are implemented for the discovery page and the user can swipe/tap to switch between tabs.
- [x] Each tab is routed but can lead to dead links or pages
- [x] Documentation present where needed
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=533%3A6528)

**Review by JimLau49 — DISMISSED** [review]

**Review by bit172 — DISMISSED** [review]
LGTM! 🚀

- [x] The tabs are implemented for the discovery page and the user can swipe/tap to switch between tabs.
- [x] Each tab is routed but can lead to dead links or pages
- [x] Documentation present where needed
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=533%3A6528)

**Review by JimLau49 — DISMISSED** [review]

**Review by bit172 — DISMISSED** [review]

**Review by cristian-aldea — DISMISSED** [review]

**Review by JimLau49 — APPROVED** [review]

**Review by cristian-aldea — APPROVED** [review]

**Pull request merged**
